crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,487 @@
+        mut retry: RetryIter<'static>,
+    ) -> SessionState {
+        let Some(node_id) = self.partition_routing.get_node_by_partition(self.partition) else {
+            tokio::time::sleep(retry.next().unwrap_or_else(|| Duration::from_secs(1))).await;
+
+            return SessionState::Connecting { retry };
+        };
+
+        tokio::select! {
+            result = self
+                .networking
+                .get_connection(node_id, Swimlane::IngressData) => {
+                    match result {
+                        Ok(connection) => SessionState::Connected { connection },
+                        Err(ConnectError::Shutdown(_)) => SessionState::Shutdown,
+                        Err(err) => {
+                            debug!("Failed to connect to node {node_id}: {err}");
+                            tokio::time::sleep(retry.next().unwrap_or_else(|| Duration::from_secs(1))).await;
+                            SessionState::Connecting { retry }
+                        }
+                    }
+            }
+            _ = cancellation.cancelled() => {
+                SessionState::Shutdown
+            }
+        }
+    }
+
+    /// Re-sends all inflight batches after a connection is restored.
+    async fn replay(&mut self, connection: &Connection) -> Result<(), ConnectionClosed> {
+        // todo(azmy): to avoid all the inflight batches again and waste traffic
+        //  maybe test the connection first by sending an empty batch and wait for response
+        //  before proceeding?
+
+        for batch in self.inflight.iter_mut() {
+            let Some(permit) = connection.reserve().await else {
+                return Err(ConnectionClosed);
+            };
+
+            // resend batch
+            let reply_rx = permit
+                .send_rpc(
+                    IngestRequest::from(Arc::clone(&batch.records)),
+                    Some(self.partition.into()),
+                )
+                .expect("encoding version to match");
+            batch.reply_rx = Some(reply_rx);
+        }
+
+        Ok(())
+    }
+
+    async fn connected(
+        &mut self,
+        cancellation: &CancellationToken,
+        connection: Connection,
+    ) -> SessionState {
+        if self.replay(&connection).await.is_err() {
+            return SessionState::Disconnected;
+        }
+
+        let chunked = if self.opts.batch_timeout != Duration::ZERO {
+            ChunksTimeout::new(&mut self.rx, self.opts.batch_size, self.opts.batch_timeout)
+                .left_stream()
+        } else {
+            (&mut self.rx)
+                .ready_chunks(self.opts.batch_size)
+                .right_stream()
+        };
+
+        let mut chunked = std::pin::pin!(chunked);
+
+        let state = loop {
+            let head: OptionFuture<_> = self
+                .inflight
+                .front_mut()
+                .and_then(|batch| batch.reply_rx.as_mut())
+                .into();
+
+            tokio::select! {
```

Comment:
Should we also await `connection.closed()` as part of the select statement to be able to react to connection losses if there is no input on `chunked`?